core/src/main/java/org/eclipse/hono/config/KeyLoader.java

```diff
@@ -90,79 +129,106 @@ public static KeyLoader fromKeyStore(final Vertx vertx, final String keyStorePat
      * @throws IllegalArgumentException if any of the files does not exist.
      */
     public static KeyLoader fromFiles(final Vertx vertx, final String keyPath, final String certPath) {
-        KeyLoader result = new KeyLoader(vertx);
-        result.loadKeysFromFiles(keyPath, certPath);
-        return result;
-    }
 
-    public PrivateKey getPrivateKey() {
-        return privateKey;
+        PrivateKey privateKey = null;
+        PublicKey publicKey = null;
+
+        if (keyPath != null) {
+            privateKey = loadPrivateKeyFromFile(vertx, keyPath);
+        }
+        if (certPath != null) {
+            publicKey = loadPublicKeyFromFile(vertx, certPath);
+        }
+
+        return new KeyLoader(privateKey, publicKey);
+
     }
 
-    public PublicKey getPublicKey() {
-        return publicKey;
+    private static PrivateKey generateRsaKey(KeySpec keySpec) throws GeneralSecurityException {
+        return KeyFactory
+                .getInstance("RSA")
+                .generatePrivate(keySpec);
     }
 
-    private void loadKeysFromFiles(final String keyPath, final String certPath) {
+    private static <R> R processFile(final Vertx vertx, final String pathName, final PemProcessor<R> processor) {
 
-        if (keyPath != null) {
-            loadPrivateKeyFromFile(keyPath);
+        final Path path = Paths.get(pathName);
+
+        if (!vertx.fileSystem().existsBlocking(pathName)) {
+            throw new IllegalArgumentException(format("%s: Private key file does not exist", path));
         }
 
-        if (certPath != null) {
-            loadPublicKeyFromFile(certPath);
+        try {
+
+            final List<Entry> pems = PemReader.readAllBlocking(vertx, path);
+
+            if (pems.isEmpty()) {
+                throw new IllegalArgumentException(format("%s: File is empty", path));
+            }
+
+            return processor.process(pems);
+
+        } catch (Exception e) {
```

Comment:
this will turn the `IllegalArgumentException` thrown above into a `RuntimeException` and consequently break the contract defined by e.g. the *fromFiles* method. I am not too fond of catching `Exception` anyway, so could we be more specific regarding what gets caught here?

Reply:
The contract should be fine, as it says "if any of the files does not exist". And the throw for this is outside of the try-catch block.

I agree that it isn't the best way to throw an catch-wrap the "empty PEM" case in the same method. My intention was to keep the current contract and still provide the information in the root cause of the wrapper exception.

My take on this would be to actually allow `processFile` (and call callers) to actually throw `Exception` and let the caller handle the problem instead explicitly. Or even narrow this down to the `GeneralSecurityException` which is the only one currently thrown in this stack.

Another option could be to add a special catch block for `RuntimeExecption`:

~~~java
catch (final RuntimeException e) {
   throw e;
}
~~~

Reply:
I'd rather extend the reasons why *fromFiles* can throw an `IllegalArgumentException`, i.e. 

`@throws IllegalArgumentException if any of the files cannot be parsed into a proper key.`

regardless of the underlying reason like *file does not exist*, *file is empty*, *file does not contain a properly encoded key*. In the end, the result is always the same and the client does not care anyway.

We can then simply always throw an `IllegalArgumentException` from within *processFile* if loading a key doesn't work.